File: Diagrammatize.Design.Pattern/code/cxx/State/A4/NightState.cpp
#include "NightState.hpp"
#include "DayState.hpp"
#include "UrgentState.hpp"
#include "Context.hpp"

NightState::NightState() {
}

State* NightState::getInstance() {
    static NightState singleton;
    return &singleton;
}

void NightState::doClock(Context* context, int hour) {
    if (9 <= hour && hour < 17) {
        context->changeState(DayState::getInstance());
    }
}

void NightState::doUse(Context* context) {
    context->callSecurityCenter("紧急：晚上使用金库！");
}

void NightState::doAlarm(Context* context) {
    context->callSecurityCenter("按下警铃(晚上)");
    context->changeState(UrgentState::getInstance());
}

void NightState::doPhone(Context* context) {
    context->recordLog("晚上的通话录音");
}

std::string NightState::toString() const {
    return "[晚上]";
}

File: Diagrammatize.Design.Pattern/code/python/State/A4/NightState.py
from State import State

class NightState(State):
    _singleton = None

    def __init__(self):
        print("create NightState")
        pass

    @classmethod
    def getInstance(cls):
        if not cls._singleton:
            cls._singleton = cls()
        return cls._singleton

    def doClock(self, context, hour):
        from DayState import DayState
        if 9 <= hour < 17:
            context.changeState(DayState.getInstance())

    def doUse(self, context):
        context.callSecurityCenter("紧急：晚上使用金库！")

    def doAlarm(self, context):
        from UrgentState import UrgentState
        context.callSecurityCenter("按下警铃(晚上)")
        context.changeState(UrgentState.getInstance())

    def doPhone(self, context):
        context.recordLog("晚上的通话录音")

    def __str__(self):
        return "[晚上]"


